### ISSUE TITLE
POST of empty LocalDateTime value causes "Failed to convert argument" JSON response

### ISSUE THREAD

**Issue by PeterFokkinga** [micronaut-projects/micronaut-core#11213]
### Expected Behavior

When posted form values are missing a value for what on the server is a LocalDateTime a NULL value should be used and the controller action should be called for the mapped url.

### Actual Behaviour

The mapped controller action is never called but Micronaut will respond with HTTP 400 Bad Request and a JSON body (even when the request has an Accept header without JSON match, but that is not the main issue)

### Steps To Reproduce

Given the following code:
```
@Controller
@Secured(SecurityRule.IS_ANONYMOUS)
public class TestController {
    @Consumes(MediaType.APPLICATION_FORM_URLENCODED)
    @Produces(MediaType.TEXT_PLAIN)
    @Post("/test")
    public HttpResponse<?> test(@Body TestBean bean) {
        try {
            verify(bean);
        } catch (ConstraintViolationException e) {
            return HttpResponse.unprocessableEntity().body(e.getMessage()); // HTTP 422
        }
        return HttpResponse.accepted().body("ok"); // HTTP 202
    }

    void verify(@Valid TestBean bean) {}
}


@Serdeable
@Introspected
public record TestBean(
        @NotBlank String text,
        @NotNull LocalDateTime ts
        )
{ }
```

HTTP request with both values:
```
POST http://localhost:9002/test
Content-Type: application/x-www-form-urlencoded

text=test&ts=2024-09-28T124012
```
--> 👍   HTTP 202 (Accepted) 

HTTP request with missing text value
```
POST http://localhost:9002/test
Content-Type: application/x-www-form-urlencoded

text=&ts=2024-09-28T124012
```
--> 👍  HTTP 422 (Unprocessable Entity)
as expected the controller action is called and the `verify` method throws a ConstraintViolationException because of the missing value

HTTP request with missing date-time value
```
POST http://localhost:9002/test
Content-Type: application/x-www-form-urlencoded

text=test&ts=
```
--> 👎 HTTP 400 (Bad Request) "Failed to convert argument [bean] for value [null] due to: Error deserializing type: TestBean bean"
the controller action is never called and therefore cannot provide the response expected by the frontend 

### Environment Information

MacOS 14.4.1
Temurin SDK 21.0.2
Micronaut 4.6.1 (can't use 4.6.2 because of micronaut-projects/micronaut-security#1808 )

### Version

4.6.1

**Comment by altro3** [comment]
You marked field ts not null. That's why you have exception

**Comment by PeterFokkinga** [comment]
@altro3 no, even without the `@NotNull` annotation in the bean I get this behaviour; the `@NotNull` should cause a ConstraintViolationException when the verify method is called but it doesn't get that far.

**Comment by dstepanov** [comment]
Try combination of Micronaut Nullable and NotNull from the validation

**Comment by PeterFokkinga** [comment]
@dstepanov that does not work either. Which is good as the String property works fine without `@Nullable`

It appears that the problem is a request with an empty value: `POST text=test&ts=` does not work, but `POST text=test` works. It is also not specific to LocalDateTime, Integer and enum properties behave the same.

More experimentation and `POST text=` results in an empty string, not null. Which may explain the above behaviour (not knowing what "empty" LocalDateTime is).

I would argue though that form fields that are left blank should be interpreted as undefined ie NULL

**Comment by sdelamo** [comment]
> I would argue though that form fields that are left blank should be interpreted as undefined ie NULL

I think we should do that.

**Comment by weesftw** [comment]
> > I would argue though that form fields that are left blank should be interpreted as undefined ie NULL
> 
> I think we should do that.

@sdelamo If we go ahead with the idea, we will end up with something like this:

| Input            | Before         | After        |
|------------------|----------------|--------------|
| z=&a=1&b=2       | [z:, a:1, b:2] | [a:1, b:2]   |
| =&a=1&b=2        | [:, a:1, b:2]  | [a:1, b:2]   |
| a=1&z&b=2        | [a:1, z:, b:2] | [a:1, b:2]   |

**Comment by yawkat** [comment]
regression test was incorrect.

**Comment by graemerocher** [comment]
Triage: valid and now covered by the rebased [#13073](https://github.com/micronaut-projects/micronaut-core/pull/13073), which supersedes #12877. The updated fix treats a single empty `application/x-www-form-urlencoded` field as absent for bean/record binding while preserving empty values for maps, objects, repeated fields, whitespace, and multipart text. The PR includes regression coverage and reports the relevant HTTP server tests passing. Keep open until merge.
